x-pack/filebeat/input/salesforce/input_test.go

```diff
@@ -0,0 +1,784 @@
+
+	expectedObjectEvent = `{"AdditionalInfo":"{}","ApiType":"N/A","ApiVersion":"N/A","Application":"salesforce_test","Browser":"Unknown","CipherSuite":"ECDHE-RSA-AES256-GCM-SHA384","City":"Mumbai","ClientVersion":"N/A","Country":"India","CountryIso":"IN","CreatedDate":"2023-12-06T05:44:34.942+0000","EvaluationTime":0,"EventDate":"2023-12-06T05:44:24.973+0000","EventIdentifier":"00044326-ed4a-421a-a0a8-e62ea626f3af","HttpMethod":"POST","Id":"000000000000000AAA","LoginGeoId":"04F5j00003NvV1cEAF","LoginHistoryId":"0Ya5j00003k2scQCAQ","LoginKey":"pgOVoLbV96U9o08W","LoginLatitude":19.0748,"LoginLongitude":72.8856,"LoginType":"Remote Access 2.0","LoginUrl":"login.salesforce.com","Platform":"Unknown","PostalCode":"400070","SessionLevel":"STANDARD","SourceIp":"134.238.252.19","Status":"Success","Subdivision":"Maharashtra","TlsProtocol":"TLS 1.2","UserId":"0055j00000AT6I1AAL","UserType":"Standard","Username":"salesforceinstance@devtest.in"}`
+)
+
+func TestFormQueryWithCursor(t *testing.T) {
+	logp.TestingSetup()
+
+	mockTimeNow(time.Date(2023, time.May, 18, 12, 0, 0, 0, time.UTC))
+	t.Cleanup(resetTimeNow)
+
+	tests := map[string]struct {
+		wantErr             error
+		cursor              mapstr.M
+		defaultSOQLTemplate string
+		valueSOQLTemplate   string
+		wantQuery           string
+		initialInterval     time.Duration
+	}{
+		"valid soql templates with nil cursor": { // expect default query with LogDate > initialInterval
+			initialInterval:     60 * 24 * time.Hour, // 60 * 24h = 1440h = 60 days = 2 months
+			defaultSOQLTemplate: `SELECT Id,CreatedDate,LogDate,LogFile FROM EventLogFile WHERE EventType = 'Login' AND Logdate > [[ (formatTime (now.Add (parseDuration "-1440h")) "RFC3339") ]] ORDER BY CreatedDate ASC NULLS FIRST`,
+			valueSOQLTemplate:   "SELECT Id,CreatedDate,LogDate,LogFile FROM EventLogFile WHERE EventType = 'Login' AND CreatedDate > [[ .cursor.logdate ]] ORDER BY CreatedDate ASC NULLS FIRST",
+			wantQuery:           "SELECT Id,CreatedDate,LogDate,LogFile FROM EventLogFile WHERE EventType = 'Login' AND Logdate > 2023-03-19T12:00:00Z ORDER BY CreatedDate ASC NULLS FIRST",
+			cursor:              nil,
+		},
+		"valid soql templates with non-empty .cursor.object.logdate": { // expect value SOQL query with .cursor.object.logdate set
+			initialInterval:     60 * 24 * time.Hour, // 60 * 24h = 1440h = 60 days = 2 months
+			defaultSOQLTemplate: `SELECT Id,CreatedDate,LogDate,LogFile FROM LoginEvent WHERE EventDate > [[ (formatTime (now.Add (parseDuration "-1440h")) "RFC3339") ]]`,
+			valueSOQLTemplate:   "SELECT Id,CreatedDate,LogDate,LogFile FROM LoginEvent WHERE  CreatedDate > [[ .cursor.object.logdate ]]",
+			wantQuery:           "SELECT Id,CreatedDate,LogDate,LogFile FROM LoginEvent WHERE  CreatedDate > 2023-05-18T12:00:00Z",
+			cursor:              mapstr.M{"object": mapstr.M{"logdate": timeNow().Format(formatRFC3339Like)}},
```

Comment:
If all that the `mockTimeNow` function is doing is replace the timeNow function with the one that returns the time it receives, it would still be better to just use the `mockTime` as above and call `Format` wherever needed. This would not only avoid a global variable i.e. timeNow from being accessed by tests, which often times run concurrently(see [this](https://github.com/golang/go/blob/182bdbb1e1d62fe53f8386e7b02c7738898d373b/src/testing/testing.go#L555)) but also get rid of any cleanup that is needed to reset it.

Reply:
I know we talked about this offline. So, yes I remember now why we cannot apply your suggestion. So, as we are depending on templating, we have funcs like `now`, `parseDuration`, etc and func `now` uses the custom `timeNow` (which returns the time set by `mockTimeNow` function). If you see our tests, the SOQL templates, you will find that we are dependent on function `now` and hence `mockTimeNow` function is important.


Refer: `(*valueTpl).Unpack`'s implementation